rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+* [Memory Key Storage (*future work*)](#memory-key-storage)
+
+Since only private keys are synced between clients in this mode, there is a
+possibility of state drift for other login information.
+
+For example, if the user is running Teleport Connect as the agent, but they
+relogin with `tsh`, Teleport Connect would not know that the keys and certs
+in their shared home directory have been replaced. Depending on what Teleport
+Connect was trying to do at the time of the login, it might accidentally load
+the new certs with the old key. This would lead to a new confusing error and,
+in the best case, another prompt to re-login.
+
+Note: in signing mode, dependent clients can perform any login operations as
+usual (e.g. `tsh login`, `tsh app login`) adding or removing certs directly in
+file storage. The dependent client will not generate new private keys in file
+storage and instead continue to use the same private keys offered by the agent
+via its sign method. This will prevent the agent from becoming out of sync
+with dependent clients.
+
+#### Login sync mode (*future work*)
+
+In login sync mode, the agent will be responsible for all login information.
+Dependent clients will query it for profiles, current profiles, active certs,
+in addition to interfacing with it for signing operations. Dependent clients
+will also be able to add or remove certs through the agent when running
+commands like `tsh app login`.
+
+The login sync agent will also notify any listening clients of login state
+changes. For example, if the user logs in with `tsh login`, the agent will
+store the certs from the login and notify any listening clients that the
+login certificates have been replaced.
+
+By running all login state storage through the agent, dependent clients can
+sync their login state without racing over file storage.
+
+This mode would be very useful for Teleport Connect to avoid state drifts
+reflected in the UI. In fact, the Teleport Connect UI would act as a dependent
+client, listening through `tshd` for any login state changes to respond to.
+
+Note: this mode is a stretch goal and is not scheduled to be completed, but
+this RFD would be incomplete if it did not at least consider this future work.
+
+### Client interface
+
+In order to utilize the Teleport Key Agent, dependent clients will need to
+interface with the agent as opposed to file storage or in-memory storage. Agent
+storage will be added as a new type of storage, with a `Keystore` implementation
+for [signing mode](#signing-mode) and a `ProfileStore` and `TrustedCertsStore`
+implementation for [login sync mode](#login-sync-mode-future-work).
+
+### Security
+
+#### Local key agent
+
+For the intended use case of using Teleport Key Agent as a local key agent,
+there is not much of concern to consider:
+
+* The Teleport Key Agent acts as a proxy for normal file storage, and as such
+is protected through file permissions on the unix socket it's served on:
+`$TELEPORT_HOME/agent.sock`.
```

Comment:
What are the protections around the socket? Will any program with access to it be able to ask the agent to sign whatever it wants?